### ISSUE TITLE
Infinite loop when creating bw file

### ISSUE THREAD

**Issue by drjsanger** [dpryan79/libBigWig#18]
Hi there,

Firstly thank you again for this library, it is helping us dramatically speed up our bw file processing. 
I have used libBigWig to create a set of tools to manipulate and generate bw files. We have recently run into what seems to be an infinite loop. See https://github.com/cancerit/cgpBigWig/issues/9

Ive done some detective work (using libBigWig master branch) and it seems that the code is getting stuck here https://github.com/dpryan79/libBigWig/blob/master/bwWrite.c#L928-L934 .

Initially I though this could be due to the fact we are missing contigs (Y for example) from the input bed that are present in the .fai file. Some print statements in libBigWig showed me that the code was hitting and infinite loop at the input file line `13	19020094	115108598	4` (as seen in the cgpBigWig issue linked above https://github.com/cancerit/cgpBigWig/issues/9), so the code isn't even reaching the end of the file and therefore finding no issue with missing contigs. 

Here is the c code utilising libBigWig and throwing the error https://github.com/cancerit/cgpBigWig/blob/develop/c/bg2bw.c . I haven't been able to find anything obvious causing this infinite loop in my code, so I'm wondering if I've found an edge case in libBigWig, or perhaps that an experienced eye could tell me where I've gone wrong. For the record I have also tried using an bwAppendIntervals where the last contig retrieved from the input bed matches the current contig when looping through the bed file and this hasn't solved the issue. 

@keiranmraine has an interest in this too

**Comment by dpryan79** [comment]
That's odd. Can you privately make available the BED file that's producing this?

**Comment by drjsanger** [comment]
Bed content:
```
1	13115	11934653	4
1	11935714	11950891	2
1	11952194	16849660	4
1	16851439	17233668	3
1	17235057	120902392	4
1	120904829	121147880	6
1	121153365	249239902	4
2	10586	87473799	4
2	87475462	87608667	3
2	87615478	243187487	4
3	60196	10186447	2
3	10187633	10298890	0
3	10300030	75426855	2
3	75428270	75645376	4
3	75646494	151508116	2
3	151509698	151550421	4
3	151551690	197908615	2
4	11960	191043877	4
5	11881	180790901	4
6	100115	171049557	4
7	24219	74060455	4
7	74061511	76136389	3
7	76137923	159127076	4
8	11773	146303788	4
9	10468	261190	5
9	262206	141146683	4
10	60968	135524334	4
11	87267	134946318	4
12	68656	133841262	4
13	19020094	115108598	4
14	19000059	107288870	4
15	20000537	20894465	5
15	20939450	102520794	4
16	60290	90159555	4
16	90160792	90292766	3
17	827	81185372	4
18	16443	78017073	4
19	226775	59118779	4
20	61097	62962869	4
21	9411409	48119634	4
22	16051248	51240820	4
X	2699554	88453891	4
X	88455130	92139938	3
X	92143669	154929412	4
```

**Comment by dpryan79** [comment]
Thanks, is GRCh38?

**Comment by drjsanger** [comment]
Sorry,I should have said. GRCh37

**Comment by dpryan79** [comment]
Thanks, I'll track this down.

**Comment by drjsanger** [comment]
I've just tried this with all lines leading up to and including the line it was getting stuck at and it works. 
Then I tried deleting one line from the bottom of the full bed file at a time until it worked
and if I the last two lines it begins to complete successfully again.
The job is nowhere near hitting the machine memory limit (it reports a max of ~16Mb) when successfully run over the bed file with 2 fewer lines.

**Comment by dpryan79** [comment]
The cause of this is a bug in how zoom level sizes are determined. In short, if all intervals are very very large then it's possible to overflow the uint32_t that's used in bigWig files to store positions. In the `fix18` branch, I added a quick fix to ensure that this doesn't occur. I'd like to look for other places where this could happen. In the interim, please try the `fix18` branch and see if that resolves the problems you've seen (it does in my test).

**Comment by drjsanger** [comment]
Changes in the `fix18` branch have fixed it our end, thanks for tracking it down so quickly! We can use the new branch for now and I'll keep an eye out for a new release.

**Comment by dpryan79** [comment]
Cool, I added one other small change and will push out a new release today.

**Comment by dpryan79** [comment]
This is now included in the 0.3.1 release.
